The Atlas GccCheckerPlugins package as an external.

gartung opened the pull request:
Although the PR is based on the gcc7 branch it can be back ported to any branch with a compiler that supports plugins and c++11.

cmsbuild commented:
A new Pull Request was created by  @gartung (Patrick Gartung) for branch IB/CMSSW_10_2_X/gcc700.

@cmsbuild, @smuzaffar, @gudrutis, @mrodozov can you please review it and eventually sign? Thanks.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.

gartung commented:
@smuzaffar @Dr15Jones

cmsbuild commented:
Pull request #4190 was updated.

gartung commented:
@smuzaffar 

When I try to run the plugin as compiled is fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms2/lib/libchecker_gccplugins.so
   /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms2/lib/libchecker_gccplugins.so: undefined symbol: _Z19get_inner_referenceP9tree_nodePlS1_PS0_P12machine_modePiS5_S5_

If I link the plugin against libcp1plugin.so, it fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms/lib/libchecker_gccplugins.so
   /home/gartung/CMSSW_10_2_0/external/slc7_amd64_gcc630/lib/libcp1plugin.so.0: undefined symbol: cp_binding_oracle

If I link the plugin against libcc1plugin.so, it fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms4/lib/libchecker_gccplugins.so
   /home/gartung/CMSSW_10_2_0/external/slc7_amd64_gcc630/lib/libcc1plugin.so.0: undefined symbol: c_binding_oracle

gartung commented:
Nevermind. I was able to determine that set up a project area for gcc630 and I compiled the plugin with gcc700.

gartung commented:
Once the gcc-checker-plugin toolfile is configure in a project area the gcc checker can be run with 

g++ -c -o /dev/null --std=c++14 -fplugin=$GCC_CHECKER_PLUGIN -fplugin-arg-libchecker_gccplugins-checkers=all $file

gartung commented:
@Dr15Jones I took the twiki text from your email and reformatted it.
https://twiki.cern.ch/twiki/bin/view/Sandbox/PatrickGartungSandbox

gartung commented:
@davidlange6 Are you opposed to this one too?

davidlange6 commented:
not specifically - presumably if we want this for more than a quick test, we should get it put into a proper distribution (add bonus, with a makefile or similar) rather than downloading all the root/fastjet etc etc etc sources that atlas uses..

smuzaffar commented on gcc-checker-plugin-toolfile.spec:
```diff
@@ -0,0 +1,47 @@
+%post
+%{relocateConfig}etc/scram.d/*.xml
+echo "GCC_CHECKER_PLUGIN_TOOLFILE_ROOT='$CMS_INSTALL_PREFIX/%{pkgrel}'; export GCC_CHECKER_PLUGIN_TOOLFILE_ROOT" > $RPM_INSTALL_PREFIX/%{pkgrel}/etc/profile.d/init.sh
+echo "setenv GCC_CHECKER_PLUGIN_TOOLFILE_ROOT '$CMS_INSTALL_PREFIX/%{pkgrel}'" > $RPM_INSTALL_PREFIX/%{pkgrel}/etc/profile.d/init.csh
```
@gartung , I don't think you need all this logic in this spec file. Just copy the one e.g. curl-toolfile.spec and update variables.

gartung replied:
Eventually you will need the path to the plugin library.

smuzaffar commented on gcc-checker-plugin.spec:
```diff
@@ -0,0 +1,23 @@
+### RPM external gcc-checker-plugin 1.0
```
@gartung , use version 2.0.9 here and replace 2.0.9 with %{realversion} in rest of the file.

davidlange6 commented:
the trouble is that 2.0.9 has nothing to do with the version number of this specific package being built..

> On Jul 18, 2018, at 1:09 PM, Malik Shahzad Muzaffar <notifications@github.com> wrote:
> 
> @smuzaffar commented on this pull request.
> 
> In gcc-checker-plugin.spec:
> 
> > @@ -0,0 +1,23 @@
> +### RPM external gcc-checker-plugin 1.0
> 
> @gartung , use version 2.0.9 here and replace 2.0.9 with %{realversion} in rest of the file.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

smuzaffar commented:
but at least it points to a tag in gitlab ... right?

davidlange6 commented:
no. It seems to be a version of a tarball full of sources from many projects

cd atlasexternals-2.0.9-671e8f2b1413ff87772ea5dc024abeb5ef6ee441
ls
APE			CheckerGccPlugins	FastJetContrib		HEPUtils		MCUtils			SoQt
Acts			Coin3D			GPerfTools		HepMCAnalysis		MKL			XRootD
BAT			Davix			Gdb			KLFitter		PyAnalysis		dSFMT
Blas			Eigen			Geant4			Lapack			Python			dcap
Boost			FFTW			GoogleTest		Lhapdf			ROOT			lwtnn
CLHEP			FastJet			HDF5			LibXml2			Simage			yampl



> On Jul 18, 2018, at 1:23 PM, Malik Shahzad Muzaffar <notifications@github.com> wrote:
> 
> but at least it points to a tag in gitlab ... right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

gartung commented:
I created this PR for review. It should not be merged yet. I plan to create a repo under cms-externals with just CheckerGccPlugin. I also want to change the Altas references to CMS. Is there a way to mark this PR WIP or For Review. I would mark the codecompass PR the same way.

cmsbuild commented:
Pull request #4190 was updated.

gartung commented:
I put the CheckerGccPlugins source in its own repository here
https://github.com/gartung/CheckerGccPlugins
I added a Makefile and changed all of the ATLAS references to CMS. This includes the checker_macros.h header that needs to be included to use the pragma macros needed to activate the thread checkers. The header could be placed in CMSSW if that is easier.

cmsbuild commented:
Pull request #4190 was updated.

cmsbuild commented:
Pull request #4190 was updated.

gartung commented:
Mostly unmodified source now in cms-external repo
https://github.com/cms-externals/CheckerGccPlugins

cmsbuild commented:
Pull request #4190 was updated.

gartung commented:
Create gcc-analyzer-c(xx)compiler.xml so gcc checker can be run with
scram b COMPILER=gcc-analyzer

smuzaffar commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/29320/console

cmsbuild commented:
Pull request #4190 was updated.

smuzaffar commented on gcc-checker-plugin-toolfile.spec:
```diff
@@ -0,0 +1,43 @@
+      <environment name="INCLUDE"   default="$GCC_CHECKER_PLUGIN_ROOT/include"/>
+    </client>
+  </tool>
+  <runtime name="GCC_CHECKER_PLUGIN" default="$GCC_CHECKER_PLUGIN_ROOT/lib/libchecker_gccplugins.so"/>
```
do we need GCC_CHECKER_PLUGIN env variable?

gartung replied:
It is no longer needed.

smuzaffar commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/29328/console

cmsbuild commented:
Pull request #4190 was updated.

cmsbuild commented:
+1
Tested at: f24c477d4ad886bfad061f12cd486b35f224f0f9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4190/29328/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4190/29328/summary.html

The workflows 1001.0, 1000.0, 140.53, 136.85, 136.8311, 136.7611, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2889291
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2889094
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files
